### PR TITLE
Use recursively merging to update the default schemas according to user schemas

### PR DESCRIPTION
**Pull request by st3107**
I am Songsheng Tao in Billinge group. I would like to recommend that we use the recursively merging of two nested dictionary to update the schemas instead of iterating the first two levels of keys. It will provide the user the freedom to add their own yaml files to db and add their own properties for the entries in yaml files.

### CODE REVIEW

**Comment by st3107** [comment]
@scopatz @CJ-Wright @sbillinge Hello, everyone. I would like to we use the recursively merging of two nested dictionary to update the schemas instead of iterating the first two levels of keys. Here is the PR. How do you think of it?

File: regolith/main.py [hunk 1]
```diff
@@ -269,9 +283,7 @@ def main(args=None):
     if "schemas" in rc._dict:
         user_schema = copy.deepcopy(rc.schemas)
         rc.schemas = copy.deepcopy(SCHEMAS)
-        for k in user_schema:
-            for k2, v in user_schema[k].items():
-                rc.schemas[k][k2].update(v)
+        merge_nested_dicts(rc.schemas, user_schema)  # merge user schemas into the default schemas
```
Comment (sbillinge) [inline regolith/main.py]: I think we don't need the comment.  It can be deleted.

File: regolith/main.py [hunk 2]
```diff
@@ -257,6 +257,20 @@ def create_parser():
     return p
 
 
+def merge_nested_dicts(dict0: dict, dict1: dict):
```
Comment (sbillinge) [inline regolith/main.py]: maybe remove type?

File: regolith/main.py [hunk 3]
```diff
@@ -257,6 +257,20 @@ def create_parser():
...
+    """Merge dict1 into dict0 recursively."""
+    for key in dict1.keys():  # iterate keys in dict1
```
Comment (sbillinge) [inline regolith/main.py]: please remove comments, the intent should be interpretable from the code itself.

File: regolith/main.py (hunk 2 above)
Comment (sbillinge) [inline regolith/main.py]: I wonder if this should be imported from ``tools.py``?

Comment (st3107) [reply regolith/main.py]: I put it in tools.

File: regolith/main.py (hunk 3 above)
Comment (st3107) [reply regolith/main.py]: Comments are removed.

File: regolith/main.py (hunk 2 above)
Comment (st3107) [reply regolith/main.py]: done.

File: regolith/main.py (hunk 1 above)
Comment (st3107) [reply regolith/main.py]: done.

**Comment by st3107** [comment]
Hello, @sbillinge . I add six test cases for the schemas updating function. They are in test_tools.py .  The other files will be further edited after we agree on the test cases.

I have a question about the use case: Shall we allow user to change the 'type' in the default schemas? For example, user changes the 'type' from 'string' to 'anyof_type': ['string', 'integer']?

**Comment by sbillinge** [comment]
changing a type from 'string' to 'anyof_type: ['string, 'integer']' may not
be needed.  Such a change should probably be merged into the main regolith.

File: tests/test_tools.py [hunk 4]
```diff
@@ -187,3 +187,257 @@ def test_number_suffix(input, expected):
...
+                },
+            },
+        },
+    },
+}
+
+USER_SCHEMA0 = {
+    "expenses": {
+        "itemized_expenses": {
+            "type": "list",
```
Comment (sbillinge) [inline tests/test_tools.py]: delete?

**Comment by st3107** [comment]
Encounter some errors in tests. Need fix ...

**Comment by st3107** [comment]
The failing of tests in travis is not due to my code. The master branch failed in the same way as my branch so the problem lies somewhere else.

**Comment by scopatz** [comment]
Thanks @st3107! Can you help fix up the test suite please?

**Comment by sbillinge** [comment]
@chiahaoliu and I dug around a bit and it looks as if it is a problem with migrating to python 3.8, which travis has done, but we haven't locally and tests are still passing.  Cerberus is giving an AttributeError so maybe something has been deprecated.  @st3107 can you dig around and see if you can figure out what is wrong; maybe talk to Tim if you need to?

**Comment by sbillinge** [comment]
update, @chiahaoliu found that cerberus 1.3.7 won't work on Python 3.8.  Apparently cerberus 1.3.8 has been released but hasn't appeared on conda-forge yet.  So this should presumably resolve itself shortly.

**Comment by sbillinge** [comment]
@st3107 the cerberus problems are now fixed.  Please can you push a new commit to generate a new build?  Maybe run black to check for syntax problems then commit and push?

**Comment by st3107** [comment]
@scopatz @sbillinge The build passed. Could you merget the PR?

**Comment by scopatz** [comment]
Thanks!

**Comment by sbillinge** [comment]
@scopatz we want to set up CI on one of our databases and it needs the changes that are in this merged PR to work.  Would it be possible to do a minor release of regolith and push it to conda-forge for us?  Thanks so much.  There is one other PR that is ready for review that could be pulled in first, or not if it isn't ready.

Also, I should mention, that if you like I am happy to be given more rights on the regolith repo to replace CJ, but only if you are ready for that.

**Pull request merged**